rfcs/20210518-prelu-postoperation/README.md

```diff
@@ -0,0 +1,59 @@
+# Introducing PReLU post operation support in oneDNN
+
+## Motivation
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation
+function. The DIEN model uses PReLu as matmul activation.
+This gets expanded in TF graph.
+![](graph_prelu.png "Graph attention layer")
+
+It becomes a bottleneck for the Attention layer in DIEN. Currently oneDNN
+supports PReLu only as
+[standalone primitive](https://oneapi-src.github.io/oneDNN/dev_guide_prelu.html).
+
+## Proposal
+
+Extend [supported postops list](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_post_ops.html)
+by adding PReLU forward postop.
+
+``` cpp
+    /// Appends a prelu forward post-op.
+    ///
+    /// The kind of this post-op is #dnnl::primitive::kind::prelu.
+    ///
+    /// The post-op can be defined as:
+    ///
+    ///      dst[:] <- prelu(dst[:], weights[:])
+    ///      prelu:
+    ///      dst[:] <- dst[:] if dst[:] > 0
+    ///      dst[:] <- dst[:] * weights[:] if dst[:] <= 0
+    ///
+    ///
+    /// @param weights_desc Learnable parameter memory descriptor.
+    void append_prelu_forward(const memory::desc &weights_desc) {
+        error::wrap_c_api(dnnl_post_ops_append_prelu(get(), &weights_desc),
+                "could not append an prelu forward post-op");
+    }
+
+    /// Returns the parameters of a prelu forward post-op.
+    ///
+    /// @param weights_desc Output memory descriptor of a second operand.
+    void get_params_prelu(memory::desc &weights_desc) const {
+        const dnnl_memory_desc_t *data;
+        error::wrap_c_api(
+                dnnl_post_ops_get_params_prelu(get(), index, &c_alg, &data),
+                "could not get parameters of a prelu post-op");
+        weights_desc.data = *data;
+    }
+
+/// Arguments for fused prelu forward.
+#define DNNL_ARG_ATTR_POST_PRELU_WEIGHTS 10240
+```
+
+Assumptions:
+- Only one prelu post operation is allowed per primitive;
```

Comment:
is there any fundamental technical reason for this limitation? if not, it might make sense to mention that this restriction can be lifted later on if needed